### ISSUE TITLE
security: possible file disclosure via COOKIEFILE

### ISSUE THREAD

**Issue by joeyh** [solatis/haskell-network-anonymous-tor#10]
This library supports authenticating with the tor control socket by reading a cookie file and sending its content to the socket.

The tor daemon typically runs as a non-root user. If this library is running as a process owned by a different user, then this seems to allow a file disclosure: The tor daemon can request any file be used as the cookie file, including a file that the tor daemon cannot read, and this library helpfully sends it the file content.

**Comment by joeyh** [comment]
https://gitweb.torproject.org/torspec.git/tree/control-spec.txt actually describes this problem, and says that the cookie authentication is deprecated and should not be used.

So, it should be removed from this library.

**Comment by sternenseemann** [comment]
No, it is still supported by tor itself at the most recent version. May it be deprecated or not, probably for the same reasons as we should support it.

**Comment by joeyh** [comment]
Checking the size of the cookie file is not sufficient. The user could well have a 32 byte secret file. It's a good first step, but SAFECOOKIE support and refusing to serve up arbitrary files for COOKIE is the real way to fix this.

**Comment by joeyh** [comment]
Let me quote the relevant part of https://gitweb.torproject.org/torspec.git/tree/control-spec.txt

```
The COOKIE authentication method exposes the user running a
controller to an unintended information disclosure attack whenever
the controller has greater filesystem read access than the process
that it has connected to.  (Note that a controller may connect to a
process other than Tor.)  It is almost never safe to use, even if
the controller's user has explicitly specified which filename to
read an authentication cookie from.  For this reason, the COOKIE
authentication method has been deprecated and will be removed from
a future version of Tor.
```

This library should not by default, and without any warning, expose an interface that is "almost never safe to use".

**Comment by sternenseemann** [comment]
Welp, why should tor expose an interface that is "almost never safe to use"?

**Comment by sternenseemann** [comment]
You could argue for keeping it in order to prevent legacy applications from breaking.

**Comment by joeyh** [comment]
sternenseemann wrote:

> Welp, why should tor expose an interface that is "almost never safe to use"?

Presumably because enough existing clients only support that interface that
it made sense to still expose it for a while. Transitions are hard.

But there's no reason for new clients written using this library to use
the old interface.

See also: https://trac.torproject.org/projects/tor/ticket/20022

## 

see shy jo

**Comment by joeyh** [comment]
sternenseemann wrote:

> You could argue for keeping it in order to prevent legacy applications from
> breaking.

The only possible legacy application that could be broken by this
library not supporting the old interface is if this library is used with
a version of the tor daemon that does not support the newer SecureCookie
interface.

The SecureCookie interface was added to tor 0.2.2.36 in 2012.

## 

see shy jo

**Comment by sternenseemann** [comment]
Oh, so the SecureCookie system is automatically enabled by the same option as the regular cookie authentification?

**Comment by joeyh** [comment]
sternenseemann wrote:

> Oh, so the SecureCookie system is automatically enabled by the same option as
> the regular cookie authentification?

Yes, I believe so.

## 

see shy jo
